File: Stack/Decode_String.py
def decodeString(s: str) -> str:
    stack = []
    num_stack = []
    
    cur_num = 0
    cur_str = ''
    for c in s:
      if c == '[':
        stack.append(cur_str)
        num_stack.append(cur_num)
        cur_str = ''
        cur_num = 0
        continue
          
      elif c == ']':          
        prev_str = stack.pop()
        num = num_stack.pop()
        cur_str = prev_str + num*cur_str
        continue
          
      if c.isdigit():
        cur_num = cur_num*10 + int(c)
      else:
        cur_str += c

    return cur_str

decodeString(s='a2[b2[ak]]')

File: Stack/Decode_String.cs
using System;
using System.Collections.Generic;

public class Program
{
    public static string DecodeString(string s)
    {
        Stack<string> stack = new Stack<string>();
        Stack<int> numStack = new Stack<int>();

        int curNum = 0;
        string curStr = "";

        foreach (char c in s)
        {
            if (c == '[')
            {
                stack.Push(curStr);
                numStack.Push(curNum);
                curStr = "";
                curNum = 0;
                continue;
            }
            else if (c == ']')
            {
                string prevStr = stack.Pop();
                int num = numStack.Pop();
                curStr = prevStr + new string(curStr.ToCharArray(), 0, curStr.Length) + new string(curStr.ToCharArray(), 0, curStr.Length);
                continue;
            }

            if (char.IsDigit(c))
            {
                curNum = curNum * 10 + int.Parse(c.ToString());
            }
            else
            {
                curStr += c;
            }
        }

        return curStr;
    }

    public static void Main(string[] args)
    {
        string result = DecodeString("a2[b2[ak]]");
        Console.WriteLine(result);
    }
}
